tests/functional/test_fusion_volume.py

```diff
@@ -0,0 +1,676 @@
+):
+    volumes_api = purefusion.VolumesApi()
+    volumes_api.get_volume = MagicMock(side_effect=purefusion.rest.ApiException)
+    module_args.update(dict_update)
+    set_module_args(module_args)
+    # run module
+    with pytest.raises(AnsibleFailJson) as ansible_error:
+        fusion_volume.main()
+    assert ansible_error.match(expected_exception_regex)
+
+
+@patch("fusion.VolumesApi")
+def test_module_not_existent_volume_with_state_absent_not_changed(
+    volumes_api, module_args
+):
+    volumes_api = purefusion.VolumesApi()
+    volumes_api.get_volume = MagicMock(side_effect=purefusion.rest.ApiException)
+    del module_args["host_access_policies"]
+    module_args["state"] = "absent"
+    set_module_args(module_args)
+    # run module
+    with pytest.raises(AnsibleExitJson) as exception:
+        fusion_volume.main()
+    assert exception.value.args[0]["changed"] is False
+
+
+@patch("fusion.OperationsApi")
+@patch("fusion.VolumesApi")
+def test_volume_create_successfully(volumes_api, operations_api, module_args):
+    operations_api = purefusion.OperationsApi()
+    volumes_api = purefusion.VolumesApi()
+    volumes_api.get_volume = MagicMock(side_effect=purefusion.rest.ApiException)
+    volumes_api.create_volume = MagicMock(return_value=OperationMock(1))
+    operations_api.get_operation = MagicMock(return_value=SuccessfulOperationMock)
+    set_module_args(module_args)
+    # run module
+    with pytest.raises(AnsibleExitJson) as exception:
+        fusion_volume.main()
+    assert exception.value.args[0]["changed"] is True
+    volumes_api.create_volume.assert_called_once()
+    volumes_api.create_volume.assert_called_with(
+        purefusion.VolumePost(
+            size=1048576,
+            storage_class=module_args["storage_class"],
+            placement_group=module_args["placement_group"],
+            name=module_args["name"],
+            display_name=module_args["display_name"],
+            protection_policy=module_args["protection_policy"],
+        ),
+        tenant_name=module_args["tenant"],
+        tenant_space_name=module_args["tenant_space"],
+    )
+
+
+@patch("fusion.OperationsApi")
+@patch("fusion.VolumesApi")
+def test_volume_create_without_display_name_successfully(
+    volumes_api, operations_api, module_args
+):
+    del module_args["display_name"]
+    operations_api = purefusion.OperationsApi()
+    volumes_api = purefusion.VolumesApi()
+    volumes_api.get_volume = MagicMock(side_effect=purefusion.rest.ApiException)
+    volumes_api.create_volume = MagicMock(return_value=OperationMock(1))
+    operations_api.get_operation = MagicMock(return_value=SuccessfulOperationMock)
+    set_module_args(module_args)
+    # run module
+    with pytest.raises(AnsibleExitJson) as exception:
+        fusion_volume.main()
+    assert exception.value.args[0]["changed"] is True
+    volumes_api.create_volume.assert_called_once()
+    volumes_api.create_volume.assert_called_with(
```

Comment:
These two lines can be merged into one: `assert_called_once_with(...)`. The same for all other tests.